### PR TITLE
Better exception handling/throwing

### PR DESCRIPTION
**Pull request by berdon**
Adding better exception handling/throwing

- Adding support for MorseL server and client options to
    - Throw exceptions on missing hub method invocation
    - Throw exceptions on missing client method invocation
    - Throw exceptions on callback method (On(...)) invocation
    - Throw exceptions by default on invalid/mangled messages
    - Rethrow exceptions thrown during the receive loop on a caller context (for client Connections)
- Added some context and clarification in some places

### CODE REVIEW

**Comment by retslig** [comment]
Looks pretty good.

File: src/MorseL.Client/Connection.cs [hunk 1]
```diff
@@ -40,28 +62,39 @@ public class Connection
...
+            options?.Invoke(_options);
             _clientWebSocket = new WebSocketClient(uri, config, securityConfig);
             _clientWebSocket.Connected += () => Connected?.Invoke();
             _clientWebSocket.Closed += () => Closed?.Invoke(null);
-            _clientWebSocket.Error += e => Error?.Invoke(e);
+
+            // Used to track exceptions and rethrow when able
+            Error += (exception) =>
+            {
+                if (_options.RethrowUnobservedExceptions)
```
Comment (ryanwersal) [inline src/MorseL.Client/Connection.cs:79]: I like the naming of this - it definitely implies the behavior you described.

File: src/MorseL.Client/Connection.cs [hunk 2]
```diff
@@ -70,55 +103,136 @@ public Task StartAsync()
...
+                            catch (Exception exception)
+                            {
+                                await HandleUnparseableReceivedInvocationDescriptor(message.Data, exception).ConfigureAwait(false);
+                                return;
+                            }
+
                             if (invocationDescriptor == null)
                             {
-                                _logger?.LogDebug("Invocation request for unknown hub method");
+                                // Valid JSON but unparseable into a known, typed invocation descriptor (unknown method name, invalid parameters)
```
Comment (ryanwersal) [inline src/MorseL.Client/Connection.cs:119]: Is there a way to tell the difference in a relatively trivial manner?

File: src/MorseL.Sockets/MorseL.Sockets.csproj [hunk 3]
```diff
@@ -15,6 +15,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Authorization" Version="1.1.1" />
     <PackageReference Include="Microsoft.AspNetCore.WebSockets" Version="1.0.1" />
+    <PackageReference Include="Nito.AsyncEx.Tasks" Version="5.0.0-pre-02" />
```
Comment (ryanwersal) [inline src/MorseL.Sockets/MorseL.Sockets.csproj:18]: Is the preview package necessary?

File: src/MorseL/Extensions/MorseLOptions.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+﻿using Microsoft.Extensions.Options;
+using MorseL.Scaleout;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace MorseL.Extensions
```
Comment (ryanwersal) [inline src/MorseL/Extensions/MorseLOptions.cs:7]: This seems an odd namespace for this class?

File: src/MorseL/HubWebSocketHandler.cs [hunk 5]
```diff
@@ -152,27 +157,30 @@ private static bool IsHubMethod(MethodInfo methodInfo)
...
             {
-                await connection.Channel.SendMessageAsync(new Message()
-                {
-                    MessageType = MessageType.Text,
-                    Data = $"Cannot find method {invocationDescriptor.MethodName} to match inbound request for {connection.Id}"
-                }).ConfigureAwait(false);
-                return;
+                // Really strange and unlikely, valid JSON and was known but not found here
+                // Likely not possible
+                await HandleMissingReceivedInvocationDescriptor(connection, invocationDescriptor);
```
Comment (ryanwersal) [inline src/MorseL/HubWebSocketHandler.cs:183]: I definitely appreciate the explicit error handling here. Compared to the previous "just send a message" this is a substantial improvement.

**Review by ryanwersal — COMMENTED** [review]
Looks overall good to me. Glad to have a lot of improved error handling throughout.

File: src/MorseL.Client/Connection.cs (hunk 2 above)
Comment (berdon) [reply src/MorseL.Client/Connection.cs:119]: Yeah...maybe... http://www.newtonsoft.com/json/help/html/JsonSchema.htm

Comment (berdon) [reply src/MorseL.Client/Connection.cs:119]: Are you ok with me creating a ticket to simplify the schema validation so I can push these changes?

Comment (ryanwersal) [reply src/MorseL.Client/Connection.cs:119]: Go for it. Truthfully the preview package is my only real concern but I'm okay with that as well. :ship:

File: src/MorseL.Sockets/MorseL.Sockets.csproj (hunk 3 above)
Comment (berdon) [reply src/MorseL.Sockets/MorseL.Sockets.csproj:18]: Necessary for .NET Standard 1.6+

File: src/MorseL/Extensions/MorseLOptions.cs (hunk 4 above)
Comment (berdon) [reply src/MorseL/Extensions/MorseLOptions.cs:7]: This mirrors SignalR's namespace for their AppBuilder/IServiceProvider extensions. I do agree that MorseLOptions inside of MorseL.Extensions is strange. Perhaps ServiceExtensions?

**Comment by ryanwersal** [comment]
LGTM

**Comment by berdon** [comment]
**Unit Tests**
Total tests: 36. Passed: 34. Failed: 0. Skipped: 2.

**Pull request merged**